tests/validator20/validate_rich_spec_test.py

```diff
@@ -0,0 +1,69 @@
+import json
+import pytest
+
+from swagger_spec_validator.common import SwaggerValidationError
+from swagger_spec_validator.validator20 import validate_spec
+
+
+@pytest.fixture
+def swagger_spec(petstore_contents):
+    return json.loads(petstore_contents)
+
+
+def test_failure_on_duplicate_api_parameters(swagger_spec):
+    param1 = {"name": "foo", "in": "query", "type": "string"}
+    param2 = {"name": "foo", "in": "query", "type": "integer"}
+    swagger_spec['paths']['/pet']['parameters'] = [param1, param2]
+    with pytest.raises(SwaggerValidationError) as exc_info:
+        validate_spec(swagger_spec)
+    assert("Duplicate param found with (name, in): ('foo', 'query')"
+           in exc_info.value)
+
+
+def test_failure_on_duplicate_operation_parameters(swagger_spec):
+    param1 = {"name": "foo", "in": "query", "type": "string"}
+    param2 = {"name": "foo", "in": "query", "type": "integer"}
+    swagger_spec['paths']['/pet']['post']['parameters'].extend([param1, param2])
+    with pytest.raises(SwaggerValidationError) as exc_info:
+        validate_spec(swagger_spec)
+    assert("Duplicate param found with (name, in): ('foo', 'query')"
+           in exc_info.value)
+
+
+def test_failure_on_unresolvable_path_parameter(swagger_spec):
+    swagger_spec['paths']['/pet/{foo}'] = swagger_spec['paths']['/pet']
+    with pytest.raises(SwaggerValidationError) as exc_info:
+        validate_spec(swagger_spec)
+    assert("Path Parameter used is not defined: foo" in exc_info.value)
+
+
+def test_failure_on_unresolvable_ref_of_props_required_list(swagger_spec):
+    swagger_spec['definitions']['Pet']['required'].append('bla')
+    with pytest.raises(SwaggerValidationError) as exc_info:
+        validate_spec(swagger_spec)
+    assert "Required list has properties not defined: ['bla']" in exc_info.value
+
+# TODO: validate definitions of references made by $ref, commented them for now.
+"""
+def test_failure_on_unresolvable_model_reference_from_model(swagger_spec):
+    swagger_spec['definitions']['Pet']['properties']['category']['$ref'] = '_'
+    with pytest.raises(SwaggerValidationError):
+        validate_spec(swagger_spec)
```

Comment:
As discussed on irc and in this json-schema [issue](https://github.com/Julian/jsonschema/issues/196), the solution could be either our own format checker Or use a RefResolver. The issue with refResolver is that one has to first manually find all the `$ref` references and then call the RefResolver to decide whether it is valid or not. Even that becomes an issue for 'model' references. We can discuss more on this later.